<a href="https://colab.research.google.com/github/suryadevsingh23999/Image-Processing/blob/main/Machine_Learning_Mini_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Python and Machine Learning Mini-Project :

Importing Libraries :

Importing Libraries that I will be using in this project 

In [ ]:
import pandas as pd
import numpy as np

Exploring the Data :

The data we will use for this example is the Iris dataset a classical dataset in Machine Learning and Statistics. It is included in Scikit-learn in the dataset module. We can load it by calling the load_iris function :

In [ ]:
from sklearn.datasets import load_iris
iris_dataset = load_iris()

The iris object that is returned by load_iris is a Bunch object, which is very similar to a dictionary. It contains keys and values :

In [ ]:
print we("keys of iris_dataset: \n{}".format (iris_dataset.keys()))

keys of iris_dataset: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


The value of the key DESCR is a short description of the dataset. Lets look at what it contains :

In [ ]:
print(iris_dataset['DESCR'][:193] + "\n...")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, pre
...


As we can see it contains some information about the dataset itself, feel free to go through it to understand more about the dataset.
Lets move to the target_names.
The value of the key target_names is an array of strings, containing the species of flower that we want to predict i.e. 'setosa' 'versicolor' and 'virginica'.

In [ ]:
print ("Target names: {}".format(iris_dataset['target_names']))

Target names: ['setosa' 'versicolor' 'virginica']


Next in line is the feature_names.

The value of feature_names is a list of strings, giving the description of each feature it includes the 'sepal length','sepal width','petal length' and 'petal width' all in centimeter(cm)

In [ ]:
print("Feature names: \n{}".format(iris_dataset['feature_names']))

Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


The data that we are going to use is itself contained in the target and data fields. The data contains numeric measurements of sepal length, sepal width, petal length & petal width in a NumPy array:

In [ ]:
print("Type of data: {}".format(type(iris_dataset['data'])))

Type of data: <class 'numpy.ndarray'>


The rows in this data array correspond to the flowers, while the columns represent the four measurements that were taken for each flower:

In [ ]:
print("Shape of data: {}".format(iris_dataset['data'].shape))

Shape of data: (150, 4)


We see that the array contains measurements of 150 flowers remember that the individual items are called samples/data points in Machine Learning and their properties are called features. So we have 150 datapoints and 4 features. The Slope of a data array is the number of samples multipplied by the number of features. This is a convention in scikit-learning and your data will always assumed to be in this shappe. Here are the feature value for the first five samples.

In [ ]:
print("First five columns of data: \h{}".format(iris_dataset['data'][:5]))

First five columns of data: \h[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


From this data we can see that all of the first five flowers have a petal width of 0.2cm and that the first flower has the longest sepal at 5.1cm So we can make this Output in the form of a table and hence you can informed or understand that which column details represent which feature.

Lets come to the last key i.e.the target key, array contains the species of each of the flowers that were measured also as a NumPy array.

In [ ]:
print("Type of target: {}".format(type(iris_dataset['target'])))

Type of target: <class 'numpy.ndarray'>


Target is a one dimensional array and on seeing the shape we can see that it contains one entry per flower:

In [ ]:
print("Shape of target:  {}".format(iris_dataset['target'].shape))

Shape of target:  (150,)


On seeing the target key and exploring the values we can easily see that the species are encoded as integers from 0 to 2:

In [ ]:
print("Target:\n{}".format(iris_dataset['target']))

Target:
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


The meaning of the numbers are given by the iris ['target_names'] array : 0 means satosa, 1 means versicolor, and 2 means virginica.

Measuring Success : Training and Testing Data:

We want to  build a Machine Learning model from this data that can predict the species of iris for this data a new set of measurements Once the model has been built and trained we would like to test wheather the model actually works, that is wheather we should trust its predections.

In Scikit learn data is usually denoted with a capital X, while label are donted by a lowercase y .This is inspired by the standard formulation (x)=y in mathematics where x is input to a function and y is the output. Following more conventions from mathematics we use a capital X because the data is a two-dimensional arda/array and a lowercase y because the target is a one-dimensional array.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    iris_dataset['data'], iris_dataset['target'], random_state=0
)

The output of the train_test_split function is X_train, X_test, y_train, y_test, which all are NumPy arrays comtains 75% of the rows of the dataset, and X_test contains the remaining 25%

In [ ]:
print("X_train shape:{}".format(X_train.shape))
print("y_train shape:{}".format(y_train.shape))
print("X_test shape:{}".format(X_test.shape))
print("y_test shape:{}".format(y_test.shape))

X_train shape:(112, 4)
y_train shape:(112,)
X_test shape:(38, 4)
y_test shape:(38,)


Inspecting the Data :

KNeighbors Classifier :-

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn= KNeighborsClassifier(n_neighbors=1)


To build the model on the training set, we call the fit method of the knn object which takes as arguments the NumPy array X_train containing the training data and the NumPy array y_train of the correspondimg from training label

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

Making Predictions:
We can now make predictions using this model on new data for which we might not know the correct labels

In [ ]:
X_new = np.array([[5,2.9,1,0.2]])

print("X_new.shape:{}".format(X_new.shape))

X_new.shape:(1, 4)


To make prediction we call the predict method of the knn object :

In [ ]:
prediction = knn.predict(X_new)

print("Prediction:{}".format(prediction))
print("Predicted target name:{}".format(iris_dataset['target_names'][prediction]))

Prediction:[0]
Predicted target name:['setosa']


In [ ]:
y_pred = knn.predict(X_test)
print("Test set predictions:\n{}".format(y_pred))

Test set predictions:
[2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
print("Test set score:{}".format( np.mean (y_pred == y_test)))

Test set score:0.9736842105263158
